### PR TITLE
WAT-454 http throttling

### PR DESCRIPTION
**Pull request by terrainvidia**

### CODE REVIEW

File: packages/http-api/src/abstract-http-client.ts [hunk 1]
```diff
@@ -6,39 +6,44 @@ import {
...
 
 const toString = c => c.toString();
 
 export abstract class AbstractHttpClient implements IHttpClient {
     protected abstract broadcast(options: IHttpRequestMessage): void;
+    private queue: IQueue<Function>;
+    private queueRunning = false;
 
-    constructor(protected transportInstance: ITransport) {
+    constructor(protected transportInstance: ITransport, private httpThrottle: number) {
```
Comment (flops) [inline packages/http-api/src/abstract-http-client.ts]: Maybe better use options object, with httpThrottle property to solve problem possible additional properties in feature.

File: packages/test-run-controller/src/test-run-controller.ts [hunk 2]
```diff
@@ -253,7 +253,8 @@ export class TestRunController extends PluggableModule implements ITestRunContro
                 worker.execute(
                     queuedTest.test,
                     queuedTest.parameters,
-                    this.config.envParameters
+                    this.config.envParameters,
+                    this.config.httpThrottle
```
Comment (flops) [inline packages/test-run-controller/src/test-run-controller.ts]: We have `runData` property in this.config.parameters. Not sure it's a good idea to make new argument for every config parameter that we will need.

File: packages/test-worker/src/worker/worker-controller.ts [hunk 3]
```diff
@@ -46,6 +46,7 @@ export class WorkerController {
 
         let isAsync = false;
 
+        this.testAPI.setHttpThrottle(message.httpThrottle);
```
Comment (flops) [inline packages/test-worker/src/worker/worker-controller.ts]: I'd prefer as with execute send this property with test parameters

**Comment by terrainvidia** [comment]
@flops I moved httpThrottle to parameters in worker. abstractHttpClient is now using params object.
Please review again.

File: packages/api/src/test-context.ts [hunk 4]
```diff
@@ -12,7 +12,15 @@ export class TestContext {
...
 
-    public http = new HttpClient(transport);
+    public http: HttpClient;
+
+    constructor(config: any) {
+        const httpClientParams = {
+            httpThrottle: config.httpThrottle,
+        };
+        
+        this.http = new HttpClient(transport, httpClientParams);
```
Comment (johnthecat) [inline packages/api/src/test-context.ts]:
```suggestion
        this.httpClient = new HttpClient(transport, {
            httpThrottle: config.httpThrottle,
        });
```

File: packages/cli/src/commands/record.ts [hunk 5]
```diff
@@ -20,7 +20,10 @@ class RecordCommand implements ICLICommand {
         const testRunController = new TestRunController(this.config, testWorker);
         const browserProxyController = browserProxyControllerFactory(this.transport);
         const webApplicationController = new WebApplicationController(browserProxyController, this.transport);
-        const httpClient = new HttpClientLocal(this.transport);
+        const httpClientParams = {
+            httpThrottle: this.config.httpThrottle,
+        };
+        const httpClient = new HttpClientLocal(this.transport, httpClientParams);
```
Comment (johnthecat) [inline packages/cli/src/commands/record.ts]: As above

File: packages/cli/src/commands/run.ts [hunk 6]
```diff
@@ -51,7 +51,10 @@ class RunCommand implements ICLICommand {
 
         const loggerServer = new LoggerServer(this.config, this.transport, this.stdout);
         const fsReader = new FSReader();
-        const httpClient = new HttpClientLocal(this.transport);
+        const httpClientParams = {
+            httpThrottle: this.config.httpThrottle,
+        };
+        const httpClient = new HttpClientLocal(this.transport, httpClientParams);
```
Comment (johnthecat) [inline packages/cli/src/commands/run.ts]: As above

File: packages/http-api/src/abstract-http-client.ts [hunk 7]
```diff
@@ -6,39 +6,45 @@ import {
...
     IHttpResponseMessage,
     IHttpRequestMessage,
-    IHttpResponseRejectMessage
+    IHttpResponseRejectMessage,
+    IQueue,
+    HttpClientParams,
 } from '@testring/types';
 import { loggerClient } from '@testring/logger';
 import { HttpCookieJar } from './cookie-jar';
+import { Queue } from '@testring/utils';
```
Comment (johnthecat) [inline packages/http-api/src/abstract-http-client.ts]: Move import up, before local `cookie-jar` module

File: packages/http-api/src/abstract-http-client.ts [hunk 8]
```diff
@@ -6,39 +6,45 @@ import {
...
 const toString = c => c.toString();
 
 export abstract class AbstractHttpClient implements IHttpClient {
     protected abstract broadcast(options: IHttpRequestMessage): void;
+    private queue: IQueue<Function>;
+    private queueRunning = false;
 
-    constructor(protected transportInstance: ITransport) {
+    constructor(protected transportInstance: ITransport, private params: HttpClientParams) {
+        this.queue = new Queue();
```
Comment (johnthecat) [inline packages/http-api/src/abstract-http-client.ts]: Don't use common words for naming 
```suggestion
        this.requestQueue = new Queue();
```

File: packages/http-api/src/abstract-http-client.ts [hunk 9]
```diff
@@ -53,6 +59,28 @@ export abstract class AbstractHttpClient implements IHttpClient {
         return (this.isValidData(request) && request.hasOwnProperty('url'));
     }
 
+    private async runQueue(): Promise<void> {
+        if (this.queueRunning) {
+            return;
+        }
+        this.queueRunning = true;
```
Comment (johnthecat) [inline packages/http-api/src/abstract-http-client.ts]: Add empty lines between logic blocks

File: packages/http-api/src/abstract-http-client.ts [hunk 10]
```diff
@@ -53,6 +59,28 @@ export abstract class AbstractHttpClient implements IHttpClient {
...
+            const item = this.queue.shift();
+            if (item) {
+                await item();
+                await new Promise(resolve => setTimeout(resolve, this.params.httpThrottle));
+            }
+        }
+        this.queueRunning = false;
+    }
+
+    private pushQueue(requestParameters: IHttpRequest, cookieJar?: IHttpCookieJar): Promise<any> {
```
Comment (johnthecat) [inline packages/http-api/src/abstract-http-client.ts]:
```suggestion
    private pushToQueue(requestParameters: IHttpRequest, cookieJar?: IHttpCookieJar): Promise<any> {
```

File: packages/http-api/src/abstract-http-client.ts [hunk 11]
```diff
@@ -53,6 +59,28 @@ export abstract class AbstractHttpClient implements IHttpClient {
...
+        return new Promise(async (resolve, reject) => {
+            this.queue.push(() => this.sendRequest(requestParameters, cookieJar).then(resolve, reject));
```
Comment (johnthecat) [inline packages/http-api/src/abstract-http-client.ts]: Move function to separated variable

File: packages/http-api/test/http-client.spec.ts [hunk 12]
```diff
@@ -131,6 +126,70 @@ describe('HttpClient', () => {
...
+
+        await Promise.all([
+            runRequest(0),
+            runRequest(1),
+            runRequest(2),
+        ]);
+
+        chai.expect(results).to.deep.equal(responses);
+    });
+    it('should execute queued requests one by one with exact timeouts', async () => {
```
Comment (johnthecat) [inline packages/http-api/test/http-client.spec.ts]: Add empty line

File: packages/http-api/test/http-client.spec.ts [hunk 13]
```diff
@@ -131,6 +126,70 @@ describe('HttpClient', () => {
...
+                uid: data.uid,
+                response: {
+                    body: responses[data.request.body.requestId]
+                }
+            });
+        });
+
+        const results: Array<any> = [];
+
+        const runRequest = async (requestId: number) => {
```
Comment (johnthecat) [inline packages/http-api/test/http-client.spec.ts]: Add description comments, it's hard to understand what's going on

File: packages/test-run-controller/src/test-run-controller.ts [hunk 14]
```diff
@@ -164,6 +164,11 @@ export class TestRunController extends PluggableModule implements ITestRunContro
     private getQueueItemWithRunData(queueItem): IQueuedTest {
         let screenshotsEnabled = false;
         let isRetryRun = queueItem.retryCount > 0;
+        const {
+            debug = false,
```
Comment (johnthecat) [inline packages/test-run-controller/src/test-run-controller.ts]: don't use default values, they must be already defined in config

**Pull request merged**